### DISCUSSION TITLE
Unable to auth by OIDC with Gitlab

### DISCUSSION THREAD

**Discussion by ezzzh** [outline/outline#3299] [Self Hosting]
**To Reproduce**
Start docker-compose by [docker-compose.yml](https://app.getoutline.com/share/770a97da-13e5-401e-9f8a-37949c19f97e/doc/docker-7pfeLP5a8t) with all needed env variables defined in docker.env (as i found in [here](https://app.getoutline.com/share/770a97da-13e5-401e-9f8a-37949c19f97e/doc/gitlab-GjNVvyv7vW)) except OIDC_USERNAME_CLAIM changed to nickname. All needed rights are delegated to gitlab application (openid, email even profile)

**Expected behavior**
User can be auth by Gitlab and pass by Outline

**Screenshots**
![изображение](https://user-images.githubusercontent.com/35534171/160476480-a5230dc4-2c87-44e0-b83a-a50e27fa81d3.png)

**Docker logs**
```
{"error":"Invalid key length","stack":"RangeError: Invalid key length\n    at Cipheriv.createCipherBase (node:internal/crypto/cipher:116:19)\n    at Cipheriv.createCipherWithIV (node:internal
/crypto/cipher:135:3)\n    at new Cipheriv (node:internal/crypto/cipher:243:3)\n    at Object.createCipheriv (node:crypto:138:10)\n    at Object.set (/opt/outline/node_modules/sequelize-encry
pted/index.js:48:33)\n    at setEncryptedColumn (/opt/outline/build/server/models/decorators/Encrypted.js:38:53)\n    at Object.set accessToken (/opt/outline/build/server/models/UserAuthentic
ation.js:61:39)\n    at Object.set (/opt/outline/node_modules/sequelize/lib/model.js:3642:32)\n    at Object.set (/opt/outline/node_modules/sequelize/lib/model.js:3623:18)\n    at Object._ini
tValues (/opt/outline/node_modules/sequelize/lib/model.js:171:10)\n    at new Model (/opt/outline/node_modules/sequelize/lib/model.js:118:10)\n    at new Model (/opt/outline/node_modules/sequ
elize-typescript/dist/model/model/model.js:14:9)\n    at new BaseModel (/opt/outline/build/server/models/base/BaseModel.js:22:5)\n    at new UserAuthentication (/opt/outline/build/server/mode
ls/UserAuthentication.js:41:5)\n    at new user_authentication (/opt/outline/build/server/models/decorators/Fix.js:18:7)\n    at Function.build (/opt/outline/node_modules/sequelize/lib/model.
js:2167:12)\n    at Function.Model.<computed> (/opt/outline/node_modules/sequelize-typescript/dist/model/model/model.js:128:28)\n    at Function.Model.<computed> [as build] (/opt/outline/node
_modules/sequelize-typescript/dist/model/model/model.js:138:28)\n    at /opt/outline/node_modules/sequelize/lib/model.js:2185:41\n    at Array.map (<anonymous>)\n    at Function.bulkBuild (/o
pt/outline/node_modules/sequelize/lib/model.js:2185:22)\n    at Function.Model.<computed> (/opt/outline/node_modules/sequelize-typescript/dist/model/model/model.js:128:28)","level":"error","m
essage":"Error during authentication"} 
```

**Outline (please complete the following information):**
- Install: self hosted 
- Version: docker (outlinewiki/outline:0.62.0  e42e3f2e6479)
- Os: Debian 11.3

**Desktop (please complete the following information):**
 - OS: Solus linux
 - Browser: FF
 - Version: 98.0

**Comment by almereyda** [comment]
Can you provide more details about your setup, especially also the GitLab and its version?

We have a working integration between Outline 0.62.0 and GitLab 14.7 here, so I might be able to contrast our configuration with yours, to see where a culprit lays.

We're using the Compose from https://github.com/outline/outline/discussions/3291#discussioncomment-2460649.

```
OIDC_CLIENT_ID=
OIDC_CLIENT_SECRET=
OIDC_AUTH_URI=https://example.com/oauth/authorize
OIDC_TOKEN_URI=https://example.com/oauth/token
OIDC_USERINFO_URI=https://example.com/oauth/userinfo
OIDC_USERNAME_CLAIM=nickname
OIDC_DISPLAY_NAME=Example Login
OIDC_SCOPES="openid"
```

**Reply by ezzzh** [reply]
Gitlab version - Enterprise Edition 14.10.0-pre, I'm using official https://gitlab.com version, my configuration:
```
OIDC_USERNAME_CLAIM=nickname
OIDC_DISPLAY_NAME=Gitlab
OIDC_SCOPES=openid email
OIDC_AUTH_URI=https://gitlab.com/oauth/authorize
OIDC_TOKEN_URI=https://gitlab.com/oauth/token
OIDC_USERINFO_URI=https://gitlab.com/oauth/userinfo
OIDC_CLIENT_ID=
OIDC_CLIENT_SECRET=
```

but i used to specify email in OIDC_SCOPES, without it i get following error in logs: 

`{"error":"An email field was not returned in the profile parameter, but is required.","stack":"UnauthorizedError: An email field was not returned in the profile parameter, but is required.\n    at AuthenticationError (/opt/outline/build/server/errors.js:33:34)\n    at OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:78:47)\n    at /opt/outline/node_modules/passport-oauth2/lib/strategy.js:198:24\n    at OAuth2Strategy._passportOauth.Strategy.userProfile (/opt/outline/build/server/routes/auth/providers/oidc.js:50:12)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)","level":"error","message":"Error during authentication"}`

**Reply by almereyda** [reply]
Which is weird. https://docs.gitlab.com/ee/integration/openid_connect_provider.html clearly states:

> The claims `sub`, `sub_legacy`, `email`, `email_verified` and `groups_direct` are included in the ID token. All other claims are available from the `/oauth/userinfo` endpoint used by OIDC clients.

As far as I remember the specification, the scopes requested need to be the same as the scopes set up for a token. Is the application secret correctly configured in GitLab, and had only the OpenID scope set? I've checked again in the application settings of our instance, and we only provide the `openid` scope for it, resulting in a successful authentication exchange between Outline and GitLab.

Unfortunately I cannot find something in the changelog that would explain, if anything changed on GitLab's side:

- https://gitlab-com.gitlab.io/cs-tools/gitlab-cs-tools/what-is-new-since/?tab=features&selectedCategories=Authentication+and+Authorization&minVersion=14_07

**Reply by almereyda** [reply]
Please note that I had seen a similar error with https://github.com/outline/outline/issues/2781, but that was using a custom build. Reverting to stock Outline helped circumventing it. Are these maybe the same error?

**Reply by ezzzh** [reply]
@almereyda , thanks for your answers, but as i mentioned above, I use docker-compose.yml from outline docs - so I use the official outline's docker images from hub.docker.com.

**Comment by almereyda** [comment]
@tommoor 

- [ ] After rereading the OP, and my response, I have found the link to https://app.getoutline.com/share/770a97da-13e5-401e-9f8a-37949c19f97e/doc/gitlab-GjNVvyv7vW, in which it should state to use `nickname` instead.

Also I am not perfectly sure if it is necessary to use the `email` scope, since it is working for us without it.

Wondering why this may be the case, and how #2781 managed to produce the same error back then.

**Comment by Limezy** [comment]
I don't know if it's related, but for your information : 
* When I generate OIDC secret as a random string of length 64, I get the same error `({"error":"Invalid key length","stack":"RangeError: Invalid key length\n    at Cipheriv.createCipherBase...)`
* When I generate it as `openssl rand -hex 32`, it works

**Reply by GaRaOne** [reply]
how you can change the secret in gitlab. for me it is autogenerated and not editable

**Reply by tommoor** [reply]
It is the SECRET_KEY that's incorrect here. It must be a specific length as documented in .env.sample

**Comment by GaRaOne** [comment]
same problem here with
gitlab: 14.10.3
outline: docker outline/ouline:0.64.3

```
- OIDC_CLIENT_ID=hex64
- OIDC_CLIENT_SECRET=hex64

- OIDC_SCOPES=openid email
- OIDC_USERNAME_CLAIM=nickname
- OIDC_DISPLAY_NAME=GitLab

- OIDC_AUTH_URI=https://<outlinehost>/oauth/authorize
- OIDC_TOKEN_URI=https://<outlinehost>/oauth/token
- OIDC_USERINFO_URI=https://<outlinehost>/oauth/userinfo
```


gitlab comes back with 
```
https://<outlinehost>/auth/oidc.callback?hex64&state=hex16
```

outline logs:
```
{"error":"Invalid key length","stack":"RangeError: Invalid key length\n    at Cipheriv.createCipherBase (node:internal/crypto/cipher:116:19)\n    at Cipheriv.createCipherWithIV (node:internal/crypto/cipher:135:3)\n    at new Cipheriv (node:internal/crypto/cipher:243:3)\n    at Object.createCipheriv (node:crypto:138:10)\n    at Object.set (/opt/outline/node_modules/sequelize-encrypted/index.js:48:33)\n    at setEncryptedColumn (/opt/outline/build/server/models/decorators/Encrypted.js:50:53)\n    at Object.set accessToken (/opt/outline/build/server/models/UserAuthentication.js:61:39)\n    at Object.set (/opt/outline/node_modules/sequelize/lib/model.js:3642:32)\n    at Object.set (/opt/outline/node_modules/sequelize/lib/model.js:3623:18)\n    at Object._initValues (/opt/outline/node_modules/sequelize/lib/model.js:171:10)\n    at new Model (/opt/outline/node_modules/sequelize/lib/model.js:118:10)\n    at new Model (/opt/outline/node_modules/sequelize-typescript/dist/model/model/model.js:14:9)\n    at new BaseModel (/opt/outline/build/server/models/base/BaseModel.js:22:5)\n    at new UserAuthentication (/opt/outline/build/server/models/UserAuthentication.js:41:5)\n    at new user_authentication (/opt/outline/build/server/models/decorators/Fix.js:18:7)\n    at Function.build (/opt/outline/node_modules/sequelize/lib/model.js:2167:12)\n    at Function.Model.<computed> (/opt/outline/node_modules/sequelize-typescript/dist/model/model/model.js:128:28)\n    at Function.Model.<computed> [as build] (/opt/outline/node_modules/sequelize-typescript/dist/model/model/model.js:138:28)\n    at /opt/outline/node_modules/sequelize/lib/model.js:2185:41\n    at Array.map (<anonymous>)\n    at Function.bulkBuild (/opt/outline/node_modules/sequelize/lib/model.js:2185:22)\n    at Function.Model.<computed> (/opt/outline/node_modules/sequelize-typescript/dist/model/model/model.js:128:28)","level":"error","message":"Error during authentication"}
```


edit: with hexXX i mean hex value with XX digits
for example: 9dab2bbf837974bc5482869cdc336272a9af84742b5dca8893e0abd310510eef as hex64

**Comment by o1298098** [comment]
```{"error":"The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client.","stack":"TokenError: The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client.\n    at OAuth2Strategy.parseErrorResponse (/opt/outline/node_modules/passport-oauth2/lib/strategy.js:373:12)\n    at OAuth2Strategy._createOAuthError (/opt/outline/node_modules/passport-oauth2/lib/strategy.js:420:16)\n    at /opt/outline/node_modules/passport-oauth2/lib/strategy.js:177:45\n    at /opt/outline/node_modules/oauth/lib/oauth2.js:191:18\n    at passBackControl (/opt/outline/node_modules/oauth/lib/oauth2.js:132:9)\n    at IncomingMessage.<anonymous> (/opt/outline/node_modules/oauth/lib/oauth2.js:157:7)\n    at IncomingMessage.emit (node:events:538:35)\n    at IncomingMessage.emit (node:domain:475:12)\n    at endReadableNT (node:internal/streams/readable:1345:12)\n    at processTicksAndRejections (node:internal/process/task_queues:83:21)","level":"error","message":"Error during authentication"}
2022-06-06T15:16:02.043898729Z {"error":"State not return in OAuth flow","stack":"BadRequestError: State not return in OAuth flow\n    at OAuthStateMismatchError (/opt/outline/build/server/errors.js:111:34)\n    at StateStore.verify (/opt/outline/build/server/utils/passport.js:43:61)\n    at OAuth2Strategy.authenticate (/opt/outline/node_modules/passport-oauth2/lib/strategy.js:222:26)\n    at attempt (/opt/outline/node_modules/passport/lib/middleware/authenticate.js:366:16)\n    at authenticate (/opt/outline/node_modules/passport/lib/middleware/authenticate.js:367:7)\n    at /opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:194:7\n    at new Promise (<anonymous>)\n    at /opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:193:12\n    at /opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:143:7\n    at new Promise (<anonymous>)\n    at passportAuthenticate (/opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:107:15)\n    at passportMiddleware (/opt/outline/build/server/middlewares/passport.js:68:7)\n    at dispatch (/opt/outline/node_modules/koa-compose/index.js:44:32)\n    at next (/opt/outline/node_modules/koa-compose/index.js:45:18)\n    at /opt/outline/node_modules/koa-router/lib/router.js:321:16\n    at dispatch (/opt/outline/node_modules/koa-compose/index.js:44:32)\n    at next (/opt/outline/node_modules/koa-compose/index.js:45:18)\n    at methodOverrideMiddleware (/opt/outline/build/server/middlewares/methodOverride.js:22:12)\n    at dispatch (/opt/outline/node_modules/koa-compose/index.js:44:32)\n    at next (/opt/outline/node_modules/koa-compose/index.js:45:18)\n    at /opt/outline/node_modules/koa-router/lib/router.js:321:16\n    at dispatch (/opt/outline/node_modules/koa-compose/index.js:44:32)","level":"error","message":"Error during authentication"}
```
Any idea about this error?

**Reply by tommoor** [reply]
The OIDC provider did not return the `state` parameter that was passed to it… or it somehow got filtered out

**Reply by zbyte** [reply]
solution here: https://github.com/outline/outline/discussions/2089#discussioncomment-667685
